packages/next/src/client/components/router-reducer/fetch-server-response.ts

```diff
@@ -183,13 +187,23 @@ export async function fetchServerResponse(
     }
 
     // Handle the `fetch` readable stream that can be unwrapped by `React.use`.
-    const [buildIdFromResponse, flightData]: NextFlightResponse =
-      await createFromFetch(Promise.resolve(res), { callServer })
+    const response: RSCPayload = await createFromFetch(Promise.resolve(res), {
+      callServer,
+    })
+
+    const buildIdFromResponse = getBuildIdFromRSCPayload(response)
+    const flightData = getFlightDataFromRSCPayload(response)
 
     if (buildId !== buildIdFromResponse) {
       return doMpaNavigation(res.url)
     }
 
+    if (!flightData) {
+      throw new Error(
+        'Invariant: Expected flight data to be present in RSC payload. This is a bug in Next.js.'
```

Comment:
🤔 In `getFlightDataFromRSCPayload` you wrote "A null FlightData is valid". ... It seems that this is only true for `ActionFlightResponse`. A proper type narrowing with discriminating union types (or possibly casting in this case), and avoiding `getFlightDataFromRSCPayload` here, would remove the need for an invariant error.

Reply:
Discussed offline, but for posterity, this defensiveness is because we can't really trust the server (due to `middleware` and other factors). So we fail early with something obvious rather than later with a cryptic message